packages/core/src/api/blockManipulation/commands/replaceBlocks/replaceBlocks.ts

```diff
@@ -8,156 +8,162 @@ import type {
+  if (columnList.childCount > 2) {
+    return;
+  }
 
-  const shouldRemoveColumnList =
-    shouldRemoveColumn && columnList.childContainer!.node.childCount === 2;
+  const firstColumnBeforePos = columnListPos + 1;
+  const $firstColumnBeforePos = tr.doc.resolve(firstColumnBeforePos);
+  const firstColumn = $firstColumnBeforePos.nodeAfter;
 
-  const isFirstColumn =
-    columnList.childContainer!.node.firstChild === column.bnBlock.node;
+  const lastColumnAfterPos = columnListPos + columnList.nodeSize - 1;
+  const $lastColumnAfterPos = tr.doc.resolve(lastColumnAfterPos);
+  const lastColumn = $lastColumnAfterPos.nodeBefore;
 
-  const blockToMove = tr.doc.slice(
-    blockInfo.bnBlock.beforePos,
-    blockInfo.bnBlock.afterPos,
-    false,
-  );
+  if (!firstColumn || !lastColumn) {
+    throw new Error("Invalid columnList: does not have child node.");
+  }
 
-  /*
-  There are 3 different cases:
-  a) remove entire column list (if no columns would be remaining)
-  b) remove just a column (if no blocks inside a column would be remaining)
-  c) keep columns (if there are blocks remaining inside a column)
-
-  Each of these 3 cases has 2 sub-cases, depending on whether the backspace happens at the start of the first (most-left) column,
-  or at the start of a non-first column.
-  */
-  if (shouldRemoveColumnList) {
-    if (isFirstColumn) {
-      tr.step(
-        new ReplaceAroundStep(
-          // replace entire column list
-          columnList.bnBlock.beforePos,
-          columnList.bnBlock.afterPos,
-          // select content of remaining column:
-          column.bnBlock.afterPos + 1,
-          columnList.bnBlock.afterPos - 2,
-          blockToMove,
-          blockToMove.size, // append existing content to blockToMove
-          false,
-        ),
-      );
-      const pos = tr.doc.resolve(columnList.bnBlock.beforePos);
-      tr.setSelection(TextSelection.between(pos, pos));
-
-      return pos;
-    } else {
-      // replaces the column list with the blockToMove slice, prepended with the content of the remaining column
-      tr.step(
-        new ReplaceAroundStep(
-          // replace entire column list
-          columnList.bnBlock.beforePos,
-          columnList.bnBlock.afterPos,
-          // select content of existing column:
-          columnList.bnBlock.beforePos + 2,
-          column.bnBlock.beforePos - 1,
-          blockToMove,
-          0, // prepend existing content to blockToMove
-          false,
-        ),
-      );
-      const pos = tr.doc.resolve(tr.mapping.map(column.bnBlock.beforePos - 1));
-      tr.setSelection(TextSelection.between(pos, pos));
-
-      return pos;
-    }
-  } else if (shouldRemoveColumn) {
-    if (isFirstColumn) {
-      // delete column
-      tr.delete(column.bnBlock.beforePos, column.bnBlock.afterPos);
+  const firstColumnEmpty = isEmptyColumn(firstColumn);
+  const lastColumnEmpty = isEmptyColumn(lastColumn);
```

Comment:
I'd like to see what happens when the first & last column are the same

Reply:
I decided to throw an error in this case. Since if the columnList only has a single non-empty column, this is both an unexpected case that should never happen (PM should automatically add an empty column to make the columnList valid), and it would not be covered by the existing cases.